File: app/src/main/java/com/vn/basemvvm/ui/base/activity/BasePagerAdapter.java
package com.vn.basemvvm.ui.base.activity;

import android.util.SparseArray;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.viewpager.widget.PagerAdapter;

import java.util.LinkedList;
import java.util.Queue;


public abstract class BasePagerAdapter<VH extends BasePagerAdapter.ViewHolder>
        extends PagerAdapter {

    private final Queue<VH> cache = new LinkedList<>();
    private final SparseArray<VH> attached = new SparseArray<>();

    public abstract VH onCreateViewHolder(@NonNull ViewGroup container);

    public abstract void onBindViewHolder(@NonNull VH holder, int position);

    public void onRecycleViewHolder(@NonNull VH holder) {
    }

    public VH getViewHolder(int position) {
        return attached.get(position);
    }

    @Override
    public Object instantiateItem(ViewGroup container, int position) {
        VH holder = cache.poll();
        if (holder == null) {
            holder = onCreateViewHolder(container);
        }
        attached.put(position, holder);

        // We should not use previous layout params, since ViewPager stores
        // important information there which cannot be reused
        container.addView(holder.itemView, null);

        onBindViewHolder(holder, position);
        return holder;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        VH holder = (VH) object;
        attached.remove(position);
        container.removeView(holder.itemView);
        cache.offer(holder);
        onRecycleViewHolder(holder);
    }

    @Override
    public boolean isViewFromObject(View view, Object object) {
        ViewHolder holder = (ViewHolder) object;
        return holder.itemView == view;
    }

    @Override
    public int getItemPosition(Object object) {
        return POSITION_NONE;
    }

    public static class ViewHolder {
        public final View itemView;

        public ViewHolder(@NonNull View itemView) {
            this.itemView = itemView;
        }
    }

}

File: app/src/main/java/com/vn/basemvvm/ui/base/adapter/BasePagerAdapter.kt
package com.vn.basemvvm.ui.base.adapter

import android.util.SparseArray
import android.view.View
import android.view.ViewGroup
import androidx.viewpager.widget.PagerAdapter
import java.util.*

abstract class BasePagerAdapter<VH : BasePagerAdapter.ViewHolder?> : PagerAdapter() {
    
    private val cache: Queue<VH> = LinkedList()
    private val attached = SparseArray<VH>()
    abstract fun onCreateViewHolder(container: ViewGroup): VH
    abstract fun onBindViewHolder(holder: VH, position: Int)
    fun onRecycleViewHolder(holder: VH) {}
    
    fun getViewHolder(position: Int): VH {
        return attached[position]
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        var holder = cache.poll()
        if (holder == null) {
            holder = onCreateViewHolder(container)
        }
        attached.put(position, holder)
        container.addView(holder!!.itemView, null)
        onBindViewHolder(holder, position)
        return holder
    }

    @Suppress("UNCHECKED_CAST")
    override fun destroyItem(container: ViewGroup, position: Int, obj: Any) {
        val holder = obj as VH
        attached.remove(position)
        container.removeView(holder!!.itemView)
        cache.offer(holder)
        onRecycleViewHolder(holder)
    }

    override fun isViewFromObject(view: View, obj: Any): Boolean {
        val holder = obj as ViewHolder
        return holder.itemView === view
    }

    override fun getItemPosition(obj: Any): Int {
        return POSITION_NONE
    }

    open class ViewHolder(val itemView: View)
}